<style>
        * {
          font-family: sans-serif;
          padding-top: 5px
        }
</style>


<h1>Primer Database</h1>

<h2>Using the main page</h2>
<p>
  This is the main view for the primers, they are all displayed here.
  From top to bottom:
  <ul>
    <li>Submit buttons</li>
    <li>Filter using primer name or exact gene name</li>
    <li>Filter using a genomic position</li>
    <li>Change primer status (you can do it in bulk by checking them in the table)</li>
    <li>
      Recalculating coverage will display new coverages for the pair of primers selected.<br>
      It WILL NOT create a new pair with the primer selected
    </li>
  </ul>
  You then have a view of all the primers saved in the database with the following info displayed:
  <ul>
    <li>ID</li>
    <li>Name</li>
    <li>Gene</li>
    <li>Sequence</li>
    <li>GC %</li>
    <li>Tm</li>
    <li>Chr</li>
    <li>GRCh37 Coverage --> clickable link to UCSC</li>
    <li>GRCh38 Coverage --> clickable link to UCSC</li>
    <li>+/- (strand)</li>
    <li>Buffer</li>
    <li>PCR</li>
    <li>Program</li>
    <li>Scientist</li>
    <li>Arrival</li>
    <li>Date</li>
    <li>Storage Location</li>
    <li>Last Date Used</li>
    <li>Status</li>
    <li>SNP --> green = No SNPs detected, red = SNPs detected, blue = SNPs checked by scientist, purple = Gene not recognized in gnomAD</li>
    <li>Comments</li>
  </ul>
  Clicking on a primer will send you to the editing page of that primer.
</p>
<h2>Submitting Primers</h2>

<blockquote>
    <h3>Submitting Single Primers</h3>
      <p>
        The form has the following fields:
        <ul>
          <li>Primer name</li>
          <li>Gene name</li>
          <li>Sequence</li>
          <li>Buffer</li>
          <li>PCR Program</li>
          <li>Arrival date</li>
          <li>Storage location</li>
          <li>Comments (optional)</li>
          <li>Forename</li>
          <li>Surname</li>
        </ul>
        A few checks will be run after you submit a primer:
        <ul>
          <li>Does the primer name already exist?</li>
          <li>Does the gene name exist in our "gene file" for mapping?</li>
          <li>Does the sequence contain characters other that ATGC?</li>
        </ul>
        A mapping process + SNP checking process will be run while submitting to get coverage and possible variants in the primer sequences.
      </p>

    <h3>Submitting Pairs of Primers</h3>
    <p>
      The form is duplicated for the 2 primers paired up.<br>
      In addition to the rest, a coverage is calculated for GRCh37 and GRCh38.
    </p>
</blockquote>

<h2>Editing Primers</h2>

<blockquote>
    <h3>Editing Single Primers</h3>
    <p>
      In editing mode, you can modify the name, buffer, pcr program, arrival date, status, storage location, comments, and name of scientist.<br>
      You can also update the last date used using the "Update date used" button.<br>
      To delete a primer, you can click the "Delete primer" primer.<br><br>
      If SNPs are detected in a primer, scientists can manually clear them:
      <p>
        &nbsp;-&nbsp;A table containing the positions of the SNPs on the primer and a link to gnomAD will be present when SNPs are detected.<br>
        &nbsp;-&nbsp;Scientists can them decide for themselves whether the primer is still usable.<br>
        &nbsp;-&nbsp;A button called "SNP Check primer_name" will be present when SNPs are detected. 
        Clicking this button will change the SNP status in the index table.
      </p>
    </p>

    <h3>Editing Pairs of Primers</h3>
    <p>
      In addtion to the info in the single primer edit mode, you can delete the pair of primers
    </p>
</blockquote>